File: worker/worker.js
console.log('REDIS_URL', process.env.REDIS_URL);
console.log('JSON_RPC_API_URL', process.env.JSON_RPC_API_URL);
const HOME = process.env.HOME;

const {promisify} = require('util');

const net = require('net');
const Web3 = require('web3');
const redis = require('redis');
const assert = require('assert');
const winston = require('winston');
const bluebird = require('bluebird');

const redisClient = redis.createClient(process.env.REDIS_URL || "http://localhost:6379");
const web3 = new Web3(process.env.JSON_RPC_API_URL || "http://localhost:8545", net);

bluebird.promisifyAll(redis.Multi.prototype);

const zrangebyscoreAsync = promisify(redisClient.zrangebyscore).bind(redisClient);
// const zaddAsync = promisify(redisClient.zadd).bind(redisClient);

Promise.prototype.finally = function (cb) {
    const res = () => this;
    const fin = () => Promise.resolve(cb()).then(res);
    return this.then(fin, fin);
};

function sleep(ms = 0) {
    return new Promise(r => setTimeout(r, ms));
}

const logger = winston.createLogger({
    level: 'info',
    transports: [
        new winston.transports.File({filename: './logs/error.log', level: 'error'}),
        new winston.transports.File({filename: './logs/combined.log'})
    ]
});

if (process.env.NODE_ENV !== 'production') {
    logger.add(new winston.transports.Console({
        format: winston.format.simple()
    }));
}

async function checkConnections() {
    try {
        await web3.eth.getProtocolVersion();
    } catch (e) {
        logger.log({level: 'error', message: `Failed to connect parity node: ${e}`});
        throw new Error(`Failed to connect parity node: ${e}`);
    }

    try {
        const pingAsync = promisify(redisClient.ping).bind(redisClient);
        await pingAsync();
    } catch (e) {
        logger.log({level: 'error', message: `Failed to connect redis: ${e}`});
    }
}

// (async () => {
//     await checkConnections();
//
//     for (let i = 0; i < 100000; i++) {
//         zaddAsync('queue:blocks', 0, i);
//         if (2000 % 400) {
//             await sleep(3000);
//         }
//     }
// })();

(() => {
    redisClient.on('ready', () => {
      checkConnections()
          .then(parse)
          .catch(e => {
              logger.log({level: 'error', message: `Failed to connect parity node`});
          });
    });

    function parse() {
        zrangebyscoreAsync('queue:blocks', 0, 0)
            .then((res) => {
                let promises = [];

                let len = (res.length > 10000) ? 10000 : res.length;

                logger.log({level: 'info', message: `Blocks to parse: ${len} of ${res.length}`});

                for (let i = 0; i < len; i++) {
                    promises.push(parseBlock(res[i]));
                }

                return Promise.all(promises);
            })
            .catch(e => {
                logger.log({level: 'error', message: `Error while parsing blocks: ${e}`});
            })
            .then(() => {
                setTimeout(parse, 500);
            });
    }
})();

async function parseBlock(blockId) {
    assert.notEqual(blockId, null);

    let block;

    try {
        block = await web3.eth.getBlock(blockId);
    } catch (e) {
        logger.log({level: 'error', message: `Failed to fetch block #${blockId} information from chain: ${e}`});
        return;
    }

    if (block === null) {
      logger.log({level: 'error', message: `Null block by number ${blockId}. Skipping...`});
      return;
    }

    // multi request should go through block => txs => accounts parsing and commit changes at the end
    let multi = redisClient.multi();

    if (block && block.transactions) {
      for (let i = 0; i < block.transactions.length; i++) {
        multi.zadd('transactions:order', block.timestamp, block.transactions[i]);
      }
    }

    let blockToStore = [
        "number", block.number,
        "validator", block.author || block.miner,
        "gasUsed", block.gasUsed,
        "size", block.size,
        "timestamp", block.timestamp,
        "hash", block.hash,
    ];

    multi.hset(`block:${block.number}:detail`, ...blockToStore);

    let txHashes = block.transactions;

    if (txHashes.length !== 0) {
        logger.log({level: 'debug', message: `Block #${block.number} is empty`});

        logger.log({level: 'info', message: `Parsing block #${blockId}`});

        for (let i = 0; i < txHashes.length; i++) {
            let txHash = txHashes[i];

            await parseTransaction(multi, txHash);
            associateTxWithBlock(multi, txHash, i, blockId);
        }
    }

    multi.zadd(`queue:blocks`, 1, blockId);

    await multi.execAsync();
}

async function parseTransaction(multi, txHash) {
    assert.notEqual(multi, null);
    assert.notEqual(txHash, null);

    logger.log({level: 'info', message: `Parsing tx #${txHash}`});

    let tx = await web3.eth.getTransaction(txHash);
    let value;

    if (typeof parseInt(tx.value) === "number") {
        value = web3.utils.fromWei(tx.value);
        value = cutNumber(+value);
    } else {
        value = tx.value;
    }

    let txToStore = [
        "from", tx.from,
        "to", tx.to || "",
        "value", value,
        "gas", tx.gas,
        "gasPrice", tx.gasPrice,
        "hash", tx.hash,
        "input", tx.input,
        "nonce", tx.nonce,
        "blockNumber", tx.blockNumber,
    ];

    multi.hset(`block_tx:${tx.blockNumber}:${txHash}:detail`, ...txToStore);
    multi.zadd(`block_tx:${tx.blockNumber}:${txHash}:list`, 0, tx.from);

    // tx is a contract deployment
    if (tx.to === null) {
        // TODO: handle contract deployment
    } else {
        multi.zadd(`block_tx:${tx.blockNumber}:${txHash}:list`, 1, tx.to);
        associateAccountWithTx(multi, tx.to, tx.blockNumber, tx.hash);
        addAccountOrder(multi, tx.to, tx.blockNumber);
    }

    associateAccountWithTx(multi, tx.from, tx.blockNumber, tx.hash);
    addAccountOrder(multi, tx.from, tx.blockNumber);
}

function associateTxWithBlock(multi, txHash, order, blockId) {
    assert.notEqual(txHash, null);
    assert.notEqual(order, null);
    assert.notEqual(blockId, null);

    multi.zadd(`block:${blockId}:tx_list`, order, txHash);
}

function associateAccountWithTx(multi, accountAddress, blockNumber, txHash) {
    assert.notEqual(accountAddress, null);
    assert.notEqual(blockNumber, null);
    assert.notEqual(txHash, null);

    multi.zadd(`account:${accountAddress}:tx_list`, blockNumber, txHash);
}

function addAccountOrder(multi, accountAddress, blockNumber) {
    assert.notEqual(accountAddress, null);
    assert.notEqual(blockNumber, null);

    multi.zadd(`account:order`, blockNumber, accountAddress);
}

function cutNumber(number) {
    if (number.toString().split(".").length === 2) {
        if (number.toString().split(".")[1].length > 9) {
            return number.toFixed(9);
        }
    }
    return number;
}

File: worker/worker.go
package main

import (
	"fmt"
	"log"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
	"github.com/go-redis/redis"
	web3 "github.com/goodstemy/w3"
	"github.com/goodstemy/w3/complex/types"
	"github.com/goodstemy/w3/providers"
	"regexp"
	//abi "github.com/ethereum/go-ethereum/accounts/abi"
)

const version = "0.1.0"

var (
	redisClient   *redis.Client
	w3            *web3.Web3
	blocks        chan string
	redisURL      = os.Getenv("REDIS_URL")
	jsonRPCAPIURL = os.Getenv("JSON_RPC_API_URL")
)

func main() {
	numcpu := runtime.NumCPU()
	runtime.GOMAXPROCS(numcpu)
	blocks = make(chan string)

	urlOpts, err := redis.ParseURL(redisURL)

	if err != nil {
		log.Println("Error while getting $REDIS_URL:", err)
	}

	redisClient = redis.NewClient(&redis.Options{
		Addr:         urlOpts.Addr,
		Password:     "",
		DB:           0,
		ReadTimeout:  time.Second * 10,
		WriteTimeout: time.Second,
	})

	w3 = web3.NewWeb3(providers.NewHTTPProvider(
		strings.TrimPrefix(jsonRPCAPIURL, "http://"), 10, false))

	checkConnections()

	for i := 0; i < 10; i++ {
		go worker(blocks)
	}

	for {
		parse()
	}
}

func worker(blocks <-chan string) {
	for b := range blocks {
		parseBlock(b)
	}
}

func checkConnections() {
	_, err := redisClient.Ping().Result()

	if err != nil {
		panic(err)
	}

	_, err = w3.Eth.GetProtocolVersion()

	if err != nil {
		panic(err)
	}
}

func parse() {
	l, err := redisClient.ZCount("queue:blocks", "-inf", "+inf").Result()

	if err != nil {
		log.Printf("Error while getting zcount of queue:blocks. Error: %v\n", err)
		return
	}

	queue, err := redisClient.ZRangeByScore("queue:blocks",
		redis.ZRangeBy{
			Min:    "0",
			Max:    "0",
			Offset: int64(0),
			Count:  int64(l),
		}).Result()

	log.Printf("Zrange get %v blocks\n", len(queue))

	if err != nil {
		log.Printf("Error while getting queue:blocks. %v\n", err)
		return
	}

	var lastBlock int

	if len(queue) > 1000 {
		lastBlock = 1000
	} else {
		lastBlock = len(queue)
	}

	for _, block := range queue[:lastBlock] {
		blocks <- block
	}
}

func parseBlock(blockRaw string) {
	blockNumber, err := strconv.Atoi(blockRaw)

	if err != nil {
		log.Printf("Error while conversion %v to int. Continue...\n", blockRaw)
		return
	}

	block, err := w3.Eth.GetBlockByNumber(
		types.ComplexIntParameter(blockNumber), true)

	if err != nil {
		log.Printf("Error while getting block by number %v. Error: %v\n", blockNumber,
			err)
		return
	}

	redisClient.Pipelined(func(pipe redis.Pipeliner) error {
		for _, tx := range block.Transactions {
			_, err := pipe.ZAdd("transactions:order", redis.Z{
				Score:  float64(block.Timestamp.ToInt64()),
				Member: tx.Hash,
			}).Result()

			if err != nil {
				log.Printf("Error while adding tx:%v to transactions:order. %v\n", tx, err)
				return err
			}
		}

		addr := block.Author

		if block.Author == "" {
			addr = block.Miner
		}

		blockToStore := map[string]interface{}{
			"number":    strconv.FormatInt(block.Number.ToInt64(), 10),
			"validator": addr,
			"gasUsed":   strconv.FormatInt(block.GasUsed.ToInt64(), 10),
			"size":      strconv.FormatInt(block.Size.ToInt64(), 10),
			"timestamp": strconv.FormatInt(block.Timestamp.ToInt64(), 10),
			"hash":      block.Hash,
		}

		_, err = pipe.HMSet(
			fmt.Sprintf("block:%v:detail",
				strconv.FormatInt(block.Number.ToInt64(), 10)), blockToStore).Result()

		if err != nil {
			log.Printf("Error while HMSet block %v. Error: %v\n", blockToStore, err)
			return err
		}

		for _, tx := range block.Transactions {
			err := parseTransaction(pipe, tx.Hash)

			if err != nil {
				return err
			}
		}

		_, err = pipe.ZAdd("queue:blocks", redis.Z{
			Score:  float64(1),
			Member: strconv.FormatInt(block.Number.ToInt64(), 10),
		}).Result()

		if err != nil {
			log.Printf("Error while set score to 1 in queue:blocks on block %v\n",
				strconv.FormatInt(block.Number.ToInt64(), 10))
			return err
		}

		return nil
	})
}

func parseTransaction(pipe redis.Pipeliner, txHash string) error {
	tx, err := w3.Eth.GetTransactionByHash(txHash)

	if err != nil {
		log.Printf("Error while getting tx with hash %v. Error %v\n", txHash, err)
		return err
	}

	txToStore := map[string]interface{}{
		"from":        tx.From,
		"to":          tx.To,
		"value":       strconv.FormatInt(tx.Value.ToInt64(), 10),
		"gas":         strconv.FormatInt(tx.Gas.ToInt64(), 10),
		"gasPrice":    strconv.FormatInt(tx.GasPrice.ToInt64(), 10),
		"hash":        tx.Hash,
		"input":       tx.Input,
		"nonce":       tx.Nonce,
		"blockNumber": tx.BlockNumber,
	}

	_, err = pipe.HMSet(
		fmt.Sprintf("block_tx:%v:%v:detail", tx.BlockNumber, tx.Hash), txToStore).
		Result()

	if err != nil {
		log.Printf("Error while HMSet tx %v. Error: %v\n", txToStore, err)
		return err
	}

	_, err = pipe.ZAdd(fmt.Sprintf("block_tx:%v:%v:list", tx.BlockNumber, tx.Hash),
		redis.Z{
			Score:  0,
			Member: tx.From,
		}).Result()

	if err != nil {
		log.Printf("Error while ZAdd block_tx:list with hash %v. Error: %v\n",
			tx.Hash, err)
		return err
	}

	if tx.To == "" {
		//TODO: handle contract deployment
	} else {

		//TODO: find method in database by method prefix in transaction input
		if strings.HasPrefix(tx.Input, "0xa9059cbb") {
			//0xa9059cbb means transfer tokens method
			runes := []rune(tx.Input)
			//method := string(runes[0:10])
			re := regexp.MustCompile("^0+")
			to_address := "0x" + re.ReplaceAllLiteralString(string(runes[11:74]), "")
			value, _ := strconv.ParseUint("0x" + string(runes[75:138]), 0, 64)

			log.Printf("Token transaction in count %v to address %v\nContract: %v\nHash: %v\n",
				value, to_address, tx.To, tx.Hash)
			/*
			Output example:
			Token transaction in count 1000 to address 0xe47494379c1d48ee73454c251a6395fdd4f9eb43
			Contract: 0xadfe00d92e5a16e773891f59780e6e54f40b532e
			Hash: 0xbe5d15e63e151142a525812cc02a08f8e6489af1b7720c54378d75220fe0422b
			 */
		} else if tx.Input != "" {
			//WRONG WAY: there is no way to get abiJsonString from w3.Eth.GetCode
			//log.Printf("Unexpected input in transaction:\n%v\nTx hash: %v\nContract: %v\n",
			//	tx.Input, tx.Hash, tx.To)
			//contract_code, err := w3.Eth.GetCode(tx.To)
			//TODO: get abiJsonString from database by contract address
			//myAbi, err := abi.JSON(strings.NewReader(abiJsonString))
			//if err != nil {
			//	log.Fatal(err)
			//}
			//var ifc map[string]interface{}
			//err = myAbi.Unpack(&ifc, "someMethod", tx.Input)
			//if err != nil {
			//	log.Fatal(err)
			//}
		}
		_, err = pipe.ZAdd(fmt.Sprintf("block_tx:%v:%v:list", tx.BlockNumber, tx.Hash),
			redis.Z{
				Score:  1,
				Member: tx.To,
			}).Result()

		if err != nil {
			log.Printf("Error while ZAdd block_tx:list with hash %v. Error: %v\n",
				tx.Hash, err)
			return err
		}

		err = associateAccountWithTx(pipe, tx.To, tx.Hash, tx.BlockNumber)

		if err != nil {
			return err
		}

		err = addAccountOrder(pipe, tx.To, tx.BlockNumber)

		if err != nil {
			return err
		}
	}

	err = associateAccountWithTx(pipe, tx.From, tx.Hash, tx.BlockNumber)

	if err != nil {
		return err
	}

	err = addAccountOrder(pipe, tx.From, tx.BlockNumber)

	if err != nil {
		return err
	}

	return nil
}

func associateAccountWithTx(pipe redis.Pipeliner, accountAddress, txHash string,
	blockNumber int64) error {
	_, err := pipe.ZAdd(fmt.Sprintf("account:%v:tx_list", accountAddress),
		redis.Z{
			Score:  float64(blockNumber),
			Member: txHash,
		}).Result()

	if err != nil {
		log.Printf("Error while ZAdd block_tx:account with hash %v. Error: %v\n",
			txHash, err)
		return err
	}

	return nil
}

func addAccountOrder(pipe redis.Pipeliner, accountAddress string,
	blockNumber int64) error {
	_, err := pipe.ZAdd("account:order",
		redis.Z{
			Score:  float64(blockNumber),
			Member: accountAddress,
		}).Result()

	if err != nil {
		log.Printf("Error while ZAdd account:order with address %v. Error: %v\n",
			accountAddress, err)
		return err
	}

	return nil
}
